File: code/ch01/1.3.4.numRescueBoats.js
function binarySearch(nums, target) {
  let left = 0
  let right = nums.length - 1
  while (left <= right) {
    const mid = Math.floor(left + (right - left) / 2)
    if (nums[mid] == target) return mid
    if (nums[mid] < target)
      // 搜索区间变为 [mid+1, right]
      left = mid + 1
    if (nums[mid] > target)
      // 搜索区间变为 [left, mid - 1]
      right = mid - 1
  }
  return -1
}

File: code/ch01/1.3.4.numRescueBoats.py
def binarySearch(nums, target):
    # 左右都闭合的区间 [l, r]
    l, r = 0, len(nums) - 1
    while l <= r:
        mid = (left + right) >> 1
        if nums[mid] == target:
            return mid
        # 搜索区间变为 [mid+1, right]
        if nums[mid] < target:
            l = mid + 1
        # 搜索区间变为 [left, mid - 1]
        if nums[mid] > target:
            r = mid - 1
    return -1